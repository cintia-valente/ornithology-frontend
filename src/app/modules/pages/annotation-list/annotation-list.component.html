<app-nav-bar></app-nav-bar>
<div class="annotations">
  <div class="container">
    <div class="row justify-content-center">
      <div class="col-xl-8">
        <div class="section-title">Anotações</div>
        <table class="table table-striped">
          <thead>
            <tr>
              <th>Ave</th>
              <th>Data e Hora</th>
              <th>Local</th>
              <th>Ações</th>
            </tr>
          </thead>

          <tbody>
            <tr *ngFor="let annotation of annotations">
              <td class="py-1">
                {{ annotation.bird.namePtbr }}
              </td>
              <td class="py-1">
                {{ annotation.date | date : "dd/MM/yy HH:mm" }}
              </td>
              <td class="py-1">
                {{ annotation.place }}
              </td>

              <td class="py-1">
                <div class="table-actions">
                  <span class="icon-delete">
                    <img
                      src="assets/svg/layout/trash-solid.svg"
                      (click)="deleteAnnotation(annotation)"
                    />
                  </span>
                  <span class="icon-edit">
                    <a
                      [routerLink]="[
                        '/editar-anotacoes',
                        annotation.idAnnotation
                      ]"
                    >
                      <img
                        src="assets/svg/layout/pen-solid.svg"
                        style="width: 17px"
                      />
                    </a>
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
