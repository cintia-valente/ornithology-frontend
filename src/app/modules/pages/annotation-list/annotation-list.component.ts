import { Bird } from 'src/app/model/bird.model';
import { AnnotationService } from './../../../services/annotation.service';
import { Component } from '@angular/core';
import { Annotation } from 'src/app/model/annotation.model';
import { HttpErrorResponse } from '@angular/common/http';
import { throwError } from 'rxjs';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { ToastrService } from 'ngx-toastr';

@Component({
  selector: 'app-annotation-list',
  templateUrl: './annotation-list.component.html',
  styleUrls: ['./annotation-list.component.scss'],
})
export class AnnotationListComponent {
  annotationForm: FormGroup;
  public annotations: Annotation[] = [];
  error: boolean = false;
  birds: Bird[] = [];

  constructor(
    private formBuilder: FormBuilder,
    private annotationService: AnnotationService,
    private toastr: ToastrService
  ) {
    this.annotationForm = this.formBuilder.group({
      idAnnotation: [''],
      bird: [],
      date: [''],
      place: [
        '',
        [
          Validators.minLength(3),
          Validators.maxLength(100),
          Validators.required,
        ],
      ],
    });
  }

  ngOnInit(): void {
    this.listAnnotations();
  }

  listAnnotations() {
    this.error = false;

    this.annotationService.getAnnotations().subscribe({
      next: (data: Annotation[]) => (this.annotations = data),

      error: (err: HttpErrorResponse) => {
        this.toastr.error(
          'Erro ao carregar anotações. Por favor, tente novamente mais tarde.'
        );

        return throwError(() => err);
      },
    });
  }

  deleteAnnotation(annotation: Annotation) {
    this.annotationService
      .deleteAnnotations(annotation.idAnnotation)
      .subscribe({
        next: () => {
          this.listAnnotations();
          this.toastr.error('Excluída com sucesso.');
        },
        error: (err: HttpErrorResponse) => {
          this.toastr.error(
            'Erro ao excluir anotação. Por favor, tente novamente mais tarde.'
          );

          return throwError(() => err);
        },
      });
  }
}
