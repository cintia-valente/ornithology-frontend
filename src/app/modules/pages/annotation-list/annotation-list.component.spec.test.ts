import { HttpClientModule } from '@angular/common/http';
import {
  ComponentFixture,
  getTestBed,
  TestBed,
  TestComponentRenderer,
} from '@angular/core/testing';
import { FormBuilder, FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { Routes } from '@angular/router';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { ToastrService } from 'ngx-toastr';
import { of, throwError } from 'rxjs';

import { AnnotationService } from '../../../services/annotation.service';

import { AnnotationListComponent } from './annotation-list.component';

const mockAnnotations = [
  {
    idAnnotation: '138',
    bird: {
      id: '28',
      image: 'img',
      namePtbr: 'Canário',
      nameEnglish: 'Blue-and-yellow Macaw',
      nameLatin: 'Ara ararauna',
      size: '90 cm',
      genre: 'Can',
      color: 'Amarelo',
      family: 'Psittacidae',
      habitat: 'Floresta',
    },
    date: new Date('2022-01-01T12:00:00'),
    place: 'Porto Alegre',
  },
];

class AnnotationServiceMock {
  getAnnotations(): any {
    return of();
  }

  deleteAnnotations(id: any) {
    return of();
  }
}

class ToastrServiceMock {
  success(message?: string) {}

  error(message?: string) {}
}

const routes: Routes = [
  {
    path: 'editar-anotacoes/:id',
    component: TestComponentRenderer,
  },
];

describe('AnnotationListComponent', () => {
  let injector: TestBed;
  let component: AnnotationListComponent;
  let fixture: ComponentFixture<AnnotationListComponent>;
  let annotationService: AnnotationService;
  let toastrService: ToastrService;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        HttpClientModule,
        RouterTestingModule,
        FormsModule,
        ReactiveFormsModule,
        RouterTestingModule.withRoutes(routes),
      ],
      declarations: [AnnotationListComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        { provide: AnnotationService, useClass: AnnotationServiceMock },
        { provide: ToastrService, useClass: ToastrServiceMock },
        FormBuilder,
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(AnnotationListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  beforeEach(() => {
    injector = getTestBed();
    fixture = TestBed.createComponent(AnnotationListComponent);

    annotationService = injector.inject(AnnotationService);
    toastrService = injector.inject(ToastrService);

    jest
      .spyOn(annotationService, 'getAnnotations')
      .mockReturnValue(of(mockAnnotations));

    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it(`Dado: que o componente foi carregado
      Então: deve chamar o serviço userService.getUsers`, async () => {
    //Arrange
    const spyAnnotationList = jest
      .spyOn(annotationService, 'getAnnotations')
      .mockReturnValue(of(mockAnnotations));

    //Act
    fixture.detectChanges();

    //Assert
    expect(spyAnnotationList).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: ocorrer um erro ao carregar as anotações 
      Então: deve chamar o serviço userService.getUsers`, async () => {
    //Arrange
    const error = {
      error: 'error',
      status: 400,
      message: 'Server Error',
    };

    const spyAnnotationListError = jest
      .spyOn(annotationService, 'getAnnotations')
      .mockReturnValue(throwError(() => error.message));

    //Act
    fixture.detectChanges();

    //Assert
    expect(spyAnnotationListError).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: clicar no ícone de deletar
      Então: deve chamar o serviço userService.deleteUser`, async () => {
    //Arrange
    const mockAnnotations = {
      idAnnotation: '138',
      bird: {
        id: '28',
        image: 'img',
        namePtbr: 'Canário',
        nameEnglish: 'Blue-and-yellow Macaw',
        nameLatin: 'Ara ararauna',
        size: '90 cm',
        genre: 'Can',
        color: 'Amarelo',
        family: 'Psittacidae',
        habitat: 'Floresta',
      },
      date: new Date('2022-01-01T12:00:00'),
      place: 'Porto Alegre',
    };

    const spyUserDelete = jest
      .spyOn(annotationService, 'deleteAnnotations')
      .mockReturnValue(of(mockAnnotations));

    //Act
    fixture.detectChanges();
    await fixture.whenStable();

    const annotationDelete = fixture.nativeElement.querySelector(
      '.annotations table > tbody > tr:nth-child(1) > td:nth-child(4) > div > span.icon-delete > img'
    );

    annotationDelete.click();

    fixture.detectChanges();
    await fixture.whenStable();

    //Assert
    expect(spyUserDelete).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: atribuir um erro no serviço 
      E: clicar no ícone de deletar
      Então: deve chamar o serviço userService.deleteUser`, async () => {
    //Arrange
    const error = {
      error: 'error',
      status: 400,
      message: 'Server Error',
    };

    const spyAnnotationDeleteError = jest
      .spyOn(annotationService, 'deleteAnnotations')
      .mockReturnValue(throwError(() => error.message));

    //Act
    fixture.detectChanges();
    await fixture.whenStable();

    const annotationDelete = fixture.nativeElement.querySelector(
      '.annotations table > tbody > tr:nth-child(1) > td:nth-child(4) > div > span.icon-delete > img'
    );

    annotationDelete.click();

    fixture.detectChanges();
    await fixture.whenStable();

    //Assert
    expect(spyAnnotationDeleteError).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: detectar as alterações
      Então: deve gerar um Snapshot do componente carregado`, async () => {
    //Arrange
    jest
      .spyOn(annotationService, 'getAnnotations')
      .mockReturnValue(of(mockAnnotations));

    //Act
    fixture.detectChanges();

    //Assert
    expect(fixture).toMatchSnapshot();
  });
});
