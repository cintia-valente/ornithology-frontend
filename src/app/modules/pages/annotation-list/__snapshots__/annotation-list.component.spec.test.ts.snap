// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotationListComponent Dado: que o componente foi carregado
      Quando: detectar as alterações
      Então: deve gerar um Snapshot do componente carregado 1`] = `
<app-annotation-list
  __ngContext__={[Function Number]}
  annotationForm={[Function FormGroup]}
  annotationService={[Function AnnotationServiceMock]}
  annotations={[Function Array]}
  birds={[Function Array]}
  error="false"
  formBuilder={[Function FormBuilder]}
  toastr={[Function ToastrServiceMock]}
>
  <app-nav-bar /><div
    class="annotations"
  >
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-xl-8"
        >
          <div
            class="section-title"
          >
            Anotações
          </div>
          <table
            class="table table-striped"
          >
            <thead>
              <tr>
                <th>
                  Ave
                </th>
                <th>
                  Data e Hora
                </th>
                <th>
                  Local
                </th>
                <th>
                  Ações
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  class="py-1"
                >
                   Canário 
                </td>
                <td
                  class="py-1"
                >
                   01/01/22 12:00 
                </td>
                <td
                  class="py-1"
                >
                   Porto Alegre 
                </td>
                <td
                  class="py-1"
                >
                  <div
                    class="table-actions"
                  >
                    <span
                      class="icon-delete"
                    >
                      <img
                        src="assets/svg/layout/trash-solid.svg"
                      />
                    </span>
                    <span
                      class="icon-edit"
                    >
                      <a
                        href="/editar-anotacoes/138"
                      >
                        <img
                          src="assets/svg/layout/pen-solid.svg"
                          style="width: 17px;"
                        />
                      </a>
                    </span>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</app-annotation-list>
`;
